File: example_backend.py
from flask import Flask, request, render_template
from sqlalchemy import create_engine, Column, Integer
from sqlalchemy import String, DateTime, Text, MetaData
from sqlalchemy.orm import scoped_session, sessionmaker
from sqlalchemy.ext.declarative import declarative_base
from werkzeug import SharedDataMiddleware
from hashlib import sha1
from settings import DEBUG, AWS_ACCESS_KEY, AWS_SECRET, MIME_TYPE, BUCKET
from settings import ENGINE, PORT, CHUNK_SIZE, AWS_REGION
from urllib import quote
from datetime import datetime

import os
import hmac
import base64
import time
import json
import random
import hashlib


## The Upload DB Model
engine = create_engine(ENGINE, convert_unicode=True)
db = scoped_session(sessionmaker(
    autocommit=False, autoflush=True, bind=engine))

Base = declarative_base()
Base.query = db.query_property()
metadata = MetaData()


class Upload(Base):
    __tablename__ = 'upload'
    id = Column(Integer, primary_key=True)
    filename = Column(String(256))
    filesize = Column(String(64))
    last_modified = Column(String(64))

    upload_start = Column(DateTime)
    last_information = Column(DateTime)

    key = Column(String(256))
    upload_id = Column(String(128))
    chunks_uploaded = Column(Text)

Upload.metadata.create_all(bind=engine)


## Boilerplate
app = Flask(__name__)
app.debug = DEBUG


def init_db():
    metadata.create_all(bind=engine)
init_db()


@app.teardown_request
def teardown_db(exception=None):
    db.remove()


## Helper Functions

def _sign(key, msg):
    return hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()


def _get_signature_key(key, date_stamp, region_name, service_name):
    k_date = _sign(("AWS4" + key).encode("utf-8"), date_stamp)
    k_region = _sign(k_date, region_name)
    k_service = _sign(k_region, service_name)
    k_signing = _sign(k_service, "aws4_request")
    return k_signing


def get_signature(date):
    return _get_signature_key(
        AWS_SECRET, date.strftime("%Y%m%d"), AWS_REGION, "s3").encode('hex')

## Actual backend


@app.route("/upload-backend/signing_key/")
def signing_key():
    date = datetime.utcnow()
    key = get_signature(date)

    filename = request.args['filename']
    filesize = request.args['filesize']
    last_modified = request.args['last_modified']

    data = {
        "date": date.isoformat() + 'Z',
        "signature": key,
        "access_key": AWS_ACCESS_KEY,
        "region": AWS_REGION,
        "bucket": BUCKET,
        "backup_key": str(random.randint(1, 1000000)),
        "content_type": MIME_TYPE,
    }

    try:
        assert 'force' not in request.args
        u = db.query(Upload).filter(
            Upload.filename == filename,
            Upload.filesize == filesize,
            Upload.last_modified == last_modified
        ).first()
        assert u

        data.update({
            "key": u.key,
            "upload_id": u.upload_id,
            "chunks": map(int, u.chunks_uploaded.split(',')),
        })
    except AssertionError:
        db.query(Upload).filter(
            Upload.filename == filename,
            Upload.filesize == filesize,
            Upload.last_modified == last_modified
        ).delete()
        db.commit()

    return json.dumps(data)


@app.route("/upload-backend/chunk_loaded/")
def upload_action():
    key = request.args.get('key')
    upload_id = request.args.get('upload_id')

    filename = request.args['filename']
    filesize = request.args['filesize']
    last_modified = request.args['last_modified']
    chunk = int(request.args['chunk'])

    if filesize > CHUNK_SIZE:
        try:
            u = db.query(Upload).filter(
                Upload.filename == filename,
                Upload.filesize == filesize,
                Upload.last_modified == last_modified
            ).first()
            assert u

            chunks = set(map(int, u.chunks_uploaded.split(',')))
            chunks.add(chunk)
            u.chunks_uploaded = ','.join(map(str, chunks))
            db.commit()

        except AssertionError:
            u = Upload(
                filename=filename,
                filesize=filesize,
                last_modified=last_modified,
                chunks_uploaded=str(chunk),
                key=key,
                upload_id=upload_id,
            )
            db.add(u)
            db.commit()

    return ''


## Static files (debugging only)
@app.route("/")
def index():
    return render_template('index.html', aws_access_key=AWS_ACCESS_KEY,
                           mime_type=MIME_TYPE, bucket=BUCKET,
                           region=AWS_REGION,
                           key=str(random.randint(1, 1000000)))


if app.debug:
    app.wsgi_app = SharedDataMiddleware(app.wsgi_app, {
        '/': os.path.join(os.path.dirname(__file__), '')
    })


if __name__ == "__main__":
    app.run(host='0.0.0.0', port=PORT)

File: backends/php/example_backend.php
<?php

function getenv_default($env_var, $default) {
    $res = getenv($env_var);
    if(!$res) {
        $res = $default;
    }
    return $res;
}

class DB extends PDO {
    private static $instance = NULL;
    public static function getInstance() {
        if(!self::$instance) {
            self::$instance = new DB('sqlite:/tmp/mule-db');
            self::$instance->setup();
        }
        return self::$instance;
    }

    private function setup() {
        $this->prepare('
            CREATE TABLE IF NOT EXISTS upload (
            	id INTEGER NOT NULL,
            	filename VARCHAR(256),
            	filesize VARCHAR(64),
            	last_modified VARCHAR(64),
            	upload_start DATETIME,
            	last_information DATETIME,
            	"key" VARCHAR(256),
            	upload_id VARCHAR(128),
            	chunks_uploaded TEXT,
            	PRIMARY KEY (id)
            );
        ')->execute();
    }

    public function update_chunks($id, $chunks) {
        $query = $this->prepare(
            'UPDATE upload SET chunks_uploaded = ? WHERE id = ?'
        );
        $query->execute(array($chunks, $id));
    }

    public function get_upload($filename, $filesize, $last_modified) {
        $query = $this->prepare(
            'SELECT * FROM upload
            WHERE filename = ? AND filesize = ? AND last_modified = ?
            ORDER BY id DESC
            LIMIT 1'
        );
        $query->execute(array($filename, $filesize, $last_modified));
        $res = $query->fetchAll();
        if(count($res) == 0) {
            return NULL;
        } else {
            return $res[0];
        }
    }

    public function create_upload($filename, $filesize, $last_modified,
                                  $chunks_uploaded, $key, $upload_id) {
        $query = $this->prepare(
            'INSERT INTO upload
                (filename, filesize, last_modified,
                 chunks_uploaded, key, upload_id)
             VALUES
                (?, ?, ?, ?, ?, ?)'
        );
        $res = $query->execute(array($filename, $filesize, $last_modified,
                              $chunks_uploaded, $key, $upload_id));
    }

    public function delete_upload($filename, $filesize, $last_modified) {
        $query = $this->prepare(
            'DELETE FROM upload
            WHERE filename = ? AND filesize = ? AND last_modified = ?');
        $query->execute(array($filename, $filesize, $last_modified));
    }
}

class Backend {

    public $MIME_TYPE;
    public $BUCKET;
    public $AWS_SECRET;
    public $AWS_ACCESS_KEY;
    public $REGION;
    public $DEBUG = true;
    public $ENGINE;
    public $CHUNK_SIZE = 6291456;  // 6MB

    public function __construct() {
        $this->MIME_TYPE = getenv_default('MIME_TYPE', 'the_mime_type');
        $this->BUCKET = getenv_default('BUCKET', 'my_bucket');
        $this->AWS_SECRET = getenv_default('AWS_SECRET', 'the_secret_access_key');
        $this->AWS_ACCESS_KEY = getenv_default('AWS_ACCESS_KEY', 'the_public_access_key');
        $this->REGION = getenv_default('AWS_REGION', 'region');
        $this->ENGINE = getenv_default('DATABASE_URL', 'sqlite:database.db');
    }

    private function get_signature($date) {
        $k_date = hash_hmac(
            'sha256',$date->format("Ymd"),
            "AWS4" . $this->AWS_SECRET,
            true
        );

        $k_region = hash_hmac(
            'sha256', $this->REGION, $k_date, true
        );
        $k_service = hash_hmac(
            'sha256', "s3", $k_region, true
        );
        $k_signing = hash_hmac(
            'sha256', "aws4_request", $k_service
        );
        return $k_signing;
    }

    public function upload_action($action) {
        if(isset($_GET['key'])) {
            $key = $_GET['key'];
        }
        if(isset($_GET['upload_id'])) {
            $upload_id = $_GET['upload_id'];
        }
        if(isset($_GET['chunk'])) {
            $chunk = $_GET['chunk'];
        }
        $string = $date = NULL;

        $db = DB::getInstance();

        if($action == 'chunk_loaded') {
            $filename = $_GET['filename'];
            $filesize = (int)$_GET['filesize'];
            $last_modified = $_GET['last_modified'];
            $chunk = (int)$_GET['chunk'];

            if($filesize > $this->CHUNK_SIZE) {
                $upload = $db->get_upload($filename, $filesize,
                                          $last_modified);
                if($upload) {
                    $chunks = explode(',', $upload['chunks_uploaded']);
                    $chunks[] = $chunk;
                    $chunks = implode(',', array_unique($chunks));
                    $db->update_chunks($upload['id'], $chunks);
                } else {
                    $db->create_upload($filename, $filesize, $last_modified,
                                       $chunk, $key, $upload_id);
                }
            }

            return '';
        }

        if($action == 'signing_key') {
            $date = new DateTime('NOW');
            $date->setTimeZone(new DateTimeZone('UTC'));
            $signature = $this->get_signature($date);

            $filename = $_GET['filename'];
            $filesize = $_GET['filesize'];
            $last_modified = $_GET['last_modified'];

            $data = array(
                "date" => $date->format('c'),
                "signature" => $signature,
                "access_key" => $this->AWS_ACCESS_KEY,
                "region" => $this->REGION,
                "bucket" => $this->BUCKET,
                "backup_key" => strval(rand(0, 10000000000)),
                "content_type" => $this->MIME_TYPE
            );

            $fresh_upload = isset($_GET['force']);
            if(!$fresh_upload) {
                $upload = $db->get_upload($filename, $filesize, $last_modified);
                if($upload) {
                    $data['key'] = $upload['key'];
                    $data['upload_id'] = $upload['upload_id'];
                    $data['chunks'] = array_map(
                        'intval',
                        explode(',', $upload['chunks'])
                    );
                }
            }

            if($fresh_upload) {
                $db->delete_upload($filename, $filesize, $last_modified);
            }

            return json_encode($data);
        }
    }
}

$backend = new Backend();

if(!isset($_GET['action'])) {
    $key = (string)rand(1, 1000000);
    include 'index_template.php';
} else {
    echo $backend->upload_action($_GET['action']);
}
